Sources/SourceKitLSP/Swift/CodeActions/ConvertJSONToCodableStruct.swift

```diff
@@ -0,0 +1,318 @@
+    /// into Swift.
+    case closure(ClosureExprSyntax)
+
+    /// A closure at the end of a source file, which might have extra JSON.
+    case tail(ClosureExprSyntax, UnexpectedNodesSyntax)
+
+    /// A string literal that may contain JSON.
+    case stringLiteral(StringLiteralExprSyntax, String)
+  }
+
+  /// Look for either a closure or a string literal that might have JSON in it.
+  public static func preflightRefactoring(_ syntax: Syntax) -> Preflight? {
+    // Preflight a closure.
+    //
+    // A blob of JSON dropped into a Swift source file will look like a
+    // closure due to the curly braces. The internals might be a syntactic
+    // disaster, but we don't actually care.
+    if let closure = syntax.as(ClosureExprSyntax.self) {
+      if let file = closure.parent?.parent?.parent?.as(SourceFileSyntax.self),
+        let unexpected = file.unexpectedBetweenStatementsAndEndOfFileToken
+      {
+        return .tail(closure, unexpected)
+      }
+
+      return .closure(closure)
+    }
+
+    // We found a string literal; its contents might be JSON.
+    if let stringLit = syntax.as(StringLiteralExprSyntax.self) {
+      guard let text = stringLit.representedLiteralValue else {
+        return nil
+      }
+
+      return .stringLiteral(stringLit, text)
+    }
+
+    // Look further up the syntax tree.
+    if let parent = syntax.parent {
+      return preflightRefactoring(parent)
+    }
+
+    return nil
+  }
+}
+
+extension ConvertJSONToCodableStruct: SyntaxRefactoringCodeActionProvider {
+  static var title = "Create Codable structs from JSON"
+}
+
+extension ConvertJSONToCodableStruct {
+  fileprivate static let indentation = 4
+
+  private static func build(from jsonDictionary: Dictionary<String, Any>) -> DeclSyntax {
+    return
+      """
+      \(raw: self.buildStruct(from: jsonDictionary))
+      """
+  }
+
+  private static func buildStruct(
+    named name: String = "JSONValue",
+    at depth: Int = 0,
+    from jsonDictionary: Dictionary<String, Any>
+  ) -> String {
+    let members = self.buildStructMembers(at: depth + 1, from: jsonDictionary)
+    return
+      """
+      \(String(repeating: " ", count: depth * indentation))struct \(name): Codable {
+      \(members.joined(separator: "\n"))
+      \(String(repeating: " ", count: depth * indentation))}
+      """
+  }
+
+  private static func buildStructMembers(
+    at depth: Int,
+    from jsonDictionary: Dictionary<String, Any>
+  ) -> [String] {
+    var members: [String] = []
+
+    var uniquer: Set<String> = []
+    var nestedTypes: [(String, Dictionary<String, Any>)] = []
+    func addNestedType(_ name: String, object: Dictionary<String, Any>) {
+      guard uniquer.insert(name).inserted else {
+        return
+      }
+      nestedTypes.append((name, object))
+    }
+
+    for key in jsonDictionary.keys.sorted() {
+      guard let value = jsonDictionary[key] else {
+        continue
+      }
```

Comment:
I think this can be simplified

```suggestion
    for (key, value) in jsonDictionary.sorted(\.key) {
```